### PR TITLE
thrift/outbound: Support envelope exceptions

### PR DESCRIPTION
**Pull request by abhinav**
Enveloped responses may have the message type EXCEPTION in which case the
response body contains a Thrift struct of the shape defined by
TApplicationException in internal.thrift. We propagate that to the user as a
generic Thrift error.

### CODE REVIEW

**Comment by abhinav** [comment]
CC @yarpc/golang

File: encoding/thrift/outbound_test.go [hunk 1]
```diff
@@ -82,9 +82,14 @@ func TestClient(t *testing.T) {
...
-				Value: wire.NewValueStruct(wire.Struct{}),
+				Value: wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+					{ID: 1, Value: wire.NewValueString("great sadness")},
+					{ID: 2, Value: wire.NewValueI32(7)},
+				}}),
 			},
-			wantError: "TApplicationException: TStruct({})", // TODO: what do
+			wantError: `thrift request to procedure "MyService::someMethod" of ` +
+				`service "service" encountered an internal failure: ` +
+				"TApplicationException{Message: great sadness, Type: ProtocolError}",
```
Comment (breerly) [inline encoding/thrift/outbound_test.go]: nice 👍

**Comment by breerly** [comment]
lgtm

**Comment by prashantv** [comment]
lgtm

**Pull request merged**